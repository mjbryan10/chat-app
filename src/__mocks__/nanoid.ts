export default {
   nanoid: jest.fn(() => 'nanoid' + Math.floor(Math.random() * 10)),
};
