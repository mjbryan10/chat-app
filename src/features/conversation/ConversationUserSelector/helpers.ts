import { User } from "shared/Api/types";

/**
 * Sorts an array of users alphabetically by name.
 * @param users An array of users to sort by name
 */
export const sortUsers = (users: User[]) => {
   if (!users.length) return [];
   return users.slice().sort((a, b) => a.name.localeCompare(b.name));
};
