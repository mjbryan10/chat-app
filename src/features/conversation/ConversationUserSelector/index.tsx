import React, { useEffect } from 'react';
import UserListWithDeleteAvatar from 'features/users/UserListWithDeleteAvatar';
import UserListWithSearch from 'features/users/UserListWithSearch';
import { useSelector, useDispatch } from 'react-redux';
import {
   fetchUsers,
   selectSelectedUsers,
   clearSelectedUsers,
   clearSelectableUsers,
   selectSelectableUsers,
   addUserToSelected,
   removeUserFromSelectable,
   addUserToSelectable,
   removeUserFromSelected,
} from 'features/users/usersSlice';
import { User } from 'shared/Api/types';

const ConversationUserSelector = () => {
   //REDUX:
   const selectedUsers = useSelector(selectSelectedUsers);
   const users = useSelector(selectSelectableUsers);
   const dispatch = useDispatch();

   //LIFECYCLES:

   useEffect(() => {
      dispatch(fetchUsers());
      return () => {
         dispatch(clearSelectedUsers);
         dispatch(clearSelectableUsers);
      };
   }, [dispatch]);

   //METHODS:

   /**
    * Sorts an array of users alphabetically by name.
    * @param users An array of users to sort by name
    */
   const sortUsers = (users: User[]) => {
      //TODO: Sort alphabetically
      if (!users.length) return [];
      return users.slice().sort((a, b) => a.name.localeCompare(b.name));
   };
   /**
    * Removes a user from one array and places them in the other.
    * @param user The user from which to switch selected status
    */
   const toggleUserSelection = (user: User) => {
      const usersIndex = users.findIndex(({ id }) => user.id === id);
      if (usersIndex >= 0) {
         dispatch(addUserToSelected(user));
         dispatch(removeUserFromSelectable(user));
      } else {
         dispatch(addUserToSelectable(user));
         dispatch(removeUserFromSelected(user));
      }
   };

   return (
      <div>
         <UserListWithDeleteAvatar
            users={sortUsers(selectedUsers)}
            handleClick={toggleUserSelection}
         />
         <UserListWithSearch
            users={sortUsers(users)}
            handleItemClick={toggleUserSelection}
         />
      </div>
   );
};

export default ConversationUserSelector;
