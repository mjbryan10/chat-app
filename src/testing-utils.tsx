import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }