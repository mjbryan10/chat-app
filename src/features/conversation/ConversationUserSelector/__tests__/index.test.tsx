import React from 'react';
import { render } from 'testing-utils';
import ConversationUserSelector from '..';
import { nanoid } from 'nanoid';

jest.mock('nanoid', () => ({ nanoid: () => jest.fn() }));

describe('ConversationUserSelector Feature Component', () => {
   test('should render correctly', () => {
      const { asFragment } = render(<ConversationUserSelector />);

      expect(asFragment()).toMatchSnapshot();
   });
});
