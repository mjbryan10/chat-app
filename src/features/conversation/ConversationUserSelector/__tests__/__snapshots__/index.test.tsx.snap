// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationUserSelector Feature Component should render correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="conversation-user-selector"
  >
    <div
      class="sc-AxgMl geMXAO"
    />
    <div
      class="sc-fzoLsD jffnQx"
    >
      <label
        class="sc-fzozJi eXKbNi"
      >
        Search
        <input
          class="sc-AxmLO iKzMPF"
          data-testid="input-text-field"
          placeholder="Search for a user"
          value=""
        />
      </label>
      <ul
        class="sc-Axmtr begSdA"
      />
    </div>
  </div>
</DocumentFragment>
`;
